### PR TITLE
Extended support for filtering files copied in the exchange

### LINKED ISSUE #874: Don't allow submitting of huge files
@ellisonbg apparently had a student submit a 1TB file which filled up the drive of the server and crashed it. It seems like it would be worthwhile for `nbgrader submit` to check the file size and either exclude files that are very large or at least prompt the user "are you sure?"

### PR DESCRIPTION
**Pull request by nthiery**
- ignore too large files
- ignore non explicitly included files

Resolves #874.

### CODE REVIEW

**Comment by nthiery** [comment]
Reviews welcome.
Not fully tested yet, in particular the warnings about ignored files.
Don't merge!

File: nbgrader/utils.py [hunk 1]
```diff
@@ -215,6 +216,39 @@ def is_ignored(filename, ignore_globs=None):
             return True
     return False
 
+def ignore_patterns(exclude=None, include=None, max_file_size=None, log=None):
```
Comment (jhamrick) [inline nbgrader/utils.py]: Nice!

Would you mind adding an args and returns section to the docstring?

File: nbgrader/utils.py [hunk 2]
```diff
@@ -215,6 +216,39 @@ def is_ignored(filename, ignore_globs=None):
...
+                rationale = "excluded file"
+            else:
+                if os.path.isfile(fullname):
+                    if include and not any(fnmatch.fnmatch(filename, glob) for glob in include):
+                        rationale = "non included"
+                    elif max_file_size and os.path.getsize(fullname) > 1000*max_file_size:
+                        rationale = "file too large"
+            if rationale:
+                if log:
+                    log.warning("Ignoring {} {}".format(rationale, fullfilename))
```
Comment (jhamrick) [inline nbgrader/utils.py]: Maybe just to be clear with files that might have spaces in the name, do `"Ignoring '{}' (reason: {})".format(fullfilename, rationale)`

File: nbgrader/utils.py (hunk 1 above)
Comment (nthiery) [reply nbgrader/utils.py]: Done!

File: nbgrader/utils.py (hunk 2 above)
Comment (nthiery) [reply nbgrader/utils.py]: Done

**Comment by nthiery** [comment]
Thanks Jessica for the review! All requested changes are done now.

File: nbgrader/coursedir.py [hunk 3]
```diff
@@ -214,6 +214,29 @@ def _root_default(self):
...
+    ).tag(config=True)
+
+    include = List(
+        [
+            "*",
+        ],
+        help=dedent(
+            """
+            List of file names or file globs: non matching files
+            will be ignored with a warning.
```
Comment (jhamrick) [inline nbgrader/coursedir.py]: Maybe say you need to use the DEBUG log level to see the warning?

Comment (nthiery) [reply nbgrader/coursedir.py]: Ah, good point, forgot to update that part.

Comment (nthiery) [reply nbgrader/coursedir.py]: Done, and polished a bit the option documentation at this occasion.

Comment (jhamrick) [reply nbgrader/coursedir.py]: Thanks!

**Comment by jhamrick** [comment]
Hmm, seems that the `nbgrader/tests/utils/test_utils.py::test_ignore` test is unhappy?

**Comment by nthiery** [comment]
Oops, I indeed had forgotten to update the tests after a function name update. Fixed!

**Pull request merged**